From ffbd3ee9de02f9238547fce73a750064bcfffd4f Mon Sep 17 00:00:00 2001
From: Axel Heider <axel.heider@codasip.com>
Date: Tue, 19 Mar 2024 15:34:19 +0000
Subject: [PATCH 05/15] target/cpu: Add Codasip A730 CPU

- List all extensions from A730 manual
- Enable what QEMU supports
- Currently A730/Hobgoblin has no PMP

Co-authored-by: Stuart Menefy <stuart.menefy@codasip.com>
Signed-off-by: Axel Heider <axel.heider@codasip.com>
---
 target/riscv/cpu-qom.h      |  1 +
 target/riscv/cpu.c          | 56 +++++++++++++++++++++++++++++++++++++
 target/riscv/cpu_vendorid.h |  3 ++
 3 files changed, 60 insertions(+)

diff --git a/target/riscv/cpu-qom.h b/target/riscv/cpu-qom.h
index 91b3361dec..3b45b58c9f 100644
--- a/target/riscv/cpu-qom.h
+++ b/target/riscv/cpu-qom.h
@@ -41,6 +41,7 @@
 #define TYPE_RISCV_CPU_SIFIVE_U54       RISCV_CPU_TYPE_NAME("sifive-u54")
 #define TYPE_RISCV_CPU_THEAD_C906       RISCV_CPU_TYPE_NAME("thead-c906")
 #define TYPE_RISCV_CPU_VEYRON_V1        RISCV_CPU_TYPE_NAME("veyron-v1")
+#define TYPE_RISCV_CPU_CODASIP_A730     RISCV_CPU_TYPE_NAME("codasip-a730")
 #define TYPE_RISCV_CPU_HOST             RISCV_CPU_TYPE_NAME("host")
 
 OBJECT_DECLARE_CPU_TYPE(RISCVCPU, RISCVCPUClass, RISCV_CPU)
diff --git a/target/riscv/cpu.c b/target/riscv/cpu.c
index 83c7c0cf07..8d7902c21f 100644
--- a/target/riscv/cpu.c
+++ b/target/riscv/cpu.c
@@ -535,6 +535,61 @@ static void rv64_veyron_v1_cpu_init(Object *obj)
 #endif
 }
 
+static void rv64_codasip_a730_cpu_init(Object *obj)
+{
+    RISCVCPU *cpu = RISCV_CPU(obj);
+    CPURISCVState *env = &cpu->env;
+
+    riscv_cpu_set_misa(env, MXL_RV64, RVG | RVC | RVS | RVU);
+    env->priv_ver = PRIV_VERSION_1_12_0; /* Ss1p12 */
+
+    cpu->cfg.mvendorid = CODASIP_MVENDORID;
+    cpu->cfg.marchid = CODASIP_MARCHID;
+
+    /* A730 on Hobgoblin currently has no PMP */
+    cpu->cfg.pmp = false;
+
+    /* MMU related extensions */
+    cpu->cfg.mmu = true;
+    set_satp_mode_max_supported(cpu, VM_1_10_SV39);
+    /* Svbare */
+    /* Svade */
+
+    cpu->cfg.ext_zicsr = true;
+    /* Zicntr */
+    cpu->cfg.ext_zihpm = true;
+    cpu->cfg.pmu_mask = 0xF; /* 4 counters */
+    /* Ziccif, Ziccrse, Ziccamoa */
+    /* Za64rs */
+    /* Zicclsm */
+    cpu->cfg.ext_zihintpause = true;
+    cpu->cfg.ext_zba = true;
+    cpu->cfg.ext_zbb = true;
+    cpu->cfg.ext_zbs = true;
+    cpu->cfg.ext_zcb = true;
+    /* Zic64b */
+    cpu->cfg.ext_zicbom = true;
+    cpu->cfg.cbom_blocksize = 64;
+    /* Zicbop */
+    cpu->cfg.ext_zicboz = true;
+    cpu->cfg.cboz_blocksize = 64;
+    cpu->cfg.ext_zfhmin = true;
+    cpu->cfg.ext_zkt = true;
+    cpu->cfg.ext_zifencei = true;
+    /* Zam */
+
+    /* Sscounterenw */
+    /* Ssccptr */
+    /* Sstvecd */
+    /* Sstvala */
+    cpu->cfg.ext_svpbmt = true;
+    cpu->cfg.ext_svinval = true;
+    /* Sm1p12 */
+    /* Ssu64xl */
+    /* Sdext */
+    /* Sdtrig */
+}
+
 static void rv128_base_cpu_init(Object *obj)
 {
     if (qemu_tcg_mttcg_enabled()) {
@@ -1810,6 +1865,7 @@ static const TypeInfo riscv_cpu_type_infos[] = {
     DEFINE_CPU(TYPE_RISCV_CPU_SHAKTI_C,         rv64_sifive_u_cpu_init),
     DEFINE_CPU(TYPE_RISCV_CPU_THEAD_C906,       rv64_thead_c906_cpu_init),
     DEFINE_CPU(TYPE_RISCV_CPU_VEYRON_V1,        rv64_veyron_v1_cpu_init),
+    DEFINE_CPU(TYPE_RISCV_CPU_CODASIP_A730,     rv64_codasip_a730_cpu_init),
     DEFINE_DYNAMIC_CPU(TYPE_RISCV_CPU_BASE128,  rv128_base_cpu_init),
 #endif
 };
diff --git a/target/riscv/cpu_vendorid.h b/target/riscv/cpu_vendorid.h
index 96b6b9c2cb..34ac57d8c7 100644
--- a/target/riscv/cpu_vendorid.h
+++ b/target/riscv/cpu_vendorid.h
@@ -7,4 +7,7 @@
 #define VEYRON_V1_MIMPID        0x111
 #define VEYRON_V1_MVENDORID     0x61f
 
+#define CODASIP_MARCHID         0x8000000000010505
+#define CODASIP_MVENDORID       0x503
+
 #endif /*  TARGET_RISCV_CPU_VENDORID_H */
-- 
2.39.3

